concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java

```diff
@@ -0,0 +1,113 @@
+package kata.concurrency.testframework.api;
+
+import kata.concurrency.testframework.internal.model.InvocationContext;
+import kata.concurrency.testframework.internal.model.TestResult;
+
+import javax.annotation.PreDestroy;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.*;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class TestRunner {
+
+    private static final Logger LOGGER = Logger.getLogger(TestRunner.class.getName());
+
+    private final int maxThreadCount;
+
+    public TestRunner(int maxThreadCount) {
+        this.maxThreadCount = maxThreadCount;
+    }
+
+    public void run(Class<?>... testClasses) throws IllegalAccessException, InstantiationException,
+            ExecutionException, InterruptedException, InvocationTargetException {
+
+        ExecutorService executorService = Executors.newFixedThreadPool(maxThreadCount);
+
+        List<Object> testClassInstances = getTestClassInstances(testClasses).stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        List<InvocationContext> invocationContexts = testClassInstances.stream()
+                .flatMap(this::getInvocationContextStream)
+                .collect(Collectors.toList());
+
+        List<Future<TestResult>> futures = filterContextByAnnotation(invocationContexts, TestCase.class).stream()
```

Comment:
[question] is there any reason for not chaining the stream together? stream()...(toList) repeatedly.

Reply:
[answer] as far as I remember correctly the only reason for this was that _invocationContexts_ was being used by the _preDestroyTestClasses_ method.